docs/CONFIG.MD
# 💡 Jina AI Cloud

## ⚙️ Configurations

For deployment to Jina AI Cloud, you can configure your application infrastructure by providing `fastapi-serve` a YAML configuration file with the `--config` flag. The supported configurations are:

  - Instance type (`instance`), as defined by [Jina AI Cloud](https://docs.jina.ai/concepts/jcloud/configuration/#cpu-tiers).
  - Minimum number of replicas for your application (`autoscale_min`). Setting it 0 enables [serverless computing](https://en.wikipedia.org/wiki/Serverless_computing).
  - Disk size (`disk_size`), in GB. The default value is 1 GB.

For example:

```yaml
instance: C4
autoscale:
  min: 0
  max: 2
disk_size: 1.5G
```

You can alternatively include a `jcloud.yaml` file in your application directory with the desired configurations. However, please note that if the `--config` option is used in the command line interface, the local `jcloud.yaml` file will be disregarded. The command-line-provided configuration file will take precedence.

If you don't provide a configuration file or a specific configuration isn't specified, the following default settings will be applied: 

```yaml
instance: C3
timeout: 120
autoscale:
  min: 1
  max: 10
  metric: cpu
  target: 70
disk_size: 1G
```

## 💰 Pricing

Applications hosted on Jina AI Cloud are priced in two categories:

**Base credits**

- Base credits are what you are charged to make sure at least one instance of your application is always running and ready to handle incoming requests, ensuring high availability. If you wish to stop the application, you must either remove the app completely or put it on pause. A pause allows you to resume the app later with the same configuration. (See [`fastapi-serve` CLI section](#-fastapi-serve-cli) for more information.) Both options will halt the consumption of credits.
- Charges for base credits are calculated based on the [instance type as defined by Jina AI Cloud](https://docs.jina.ai/concepts/jcloud/configuration/#cpu-tiers).
- By default, instance type `C3` is used with a minimum of 1 instance and [Amazon EFS](https://aws.amazon.com/efs/) disk of size 1GB, which means that if your application is served on Jina AI Cloud, you will be charged ~10 credits per hour.
- You can change the instance type and the minimum number of instances by providing a YAML configuration file to the deployment with the `--config` option. For example, if you want to use instance type `C4` with a minimum of 0 replicas, and 2GB EFS disk, you can provide the following configuration file:
  ```yaml
  instance: C4
  autoscale:
    min: 0
  disk_size: 2G
  ```

**Serving credits**

- Serving credits are charged when your application is actively serving incoming requests.
- Charges for serving credits are calculated based on the credits for the instance type multiplied by the amount of time your application is serving requests. 
- You are charged for each second your application is serving requests.


**Total credits charged = Base credits + Serving credits**. ([Jina AI Cloud](https://cloud.jina.ai/pricing) defines each credit as €0.005)

### Examples

<details>
<summary><b>Example 1</b></summary>

Consider an HTTP application that has served requests for `10` minutes in the last hour and uses a custom config:
```yaml
instance: C4
autoscale:
  min: 0
disk_size: 2G
```

Total credits per hour charged would be `3.33`. The calculation is as follows:
```
C4 instance has an hourly credit rate of 20.
EFS has hourly credit rate of 0.104 per GB.
Base credits = 0 + 2 * 0.104 = 0.208 (since `autoscale_min` is 0)
Serving credits = 20 * 10/60 = 3.33
Total credits per hour = 0.208 + 3.33 = 3.538
```

</details>


<details>
<summary><b>Example 2</b></summary>

Consider a WebSocket application that had active connections for 20 minutes in the last hour and uses the default configuration.
```yaml
instance: C3
autoscale:
  min: 1
disk_size: 1G
```

Total credits per hour charged would be `13.33`. The calculation is as follows:
```
C3 instance has an hourly credit rate of 10.
EFS has hourly credit rate of 0.104 per GB.
Base credits = 10 + 1 * 0.104 = 10.104 (since `autoscale_min` is 1)
Serving credits = 10 * 20/60 = 3.33
Total credits per hour = 10.104 + 3.33 = 13.434
```

</details>


examples/CONFIG.MD
# 💡 Jina AI Cloud

## ⚙️ Configurations

For JCloud deployment, you can configure your application infrastructure by providing a YAML configuration file using the `--config` option. The supported configurations are:

  - Instance type (`instance`), as defined by [Jina AI Cloud](https://docs.jina.ai/concepts/jcloud/configuration/#cpu-tiers).
  - Minimum number of replicas for your application (`autoscale_min`). Setting it 0 enables [serverless](https://en.wikipedia.org/wiki/Serverless_computing).
  - Disk size (`disk_size`), in GB. The default value is 1 GB.

For example:

```
instance: C4
autoscale_min: 0
disk_size: 1.5G
```

You can alternatively include a `jcloud.yaml` file in your application directory with the desired configurations. However, please note that if the `--config` option is explicitly used in the command line interface, the local jcloud.yaml file will be disregarded. The command line provided configuration file will take precedence.

If you don't provide a configuration file or a specific configuration isn't specified, the following default settings will be applied: 

```
instance: C3
autoscale_min: 1
disk_size: 1G
```

## 💰 Pricing

Applications hosted on JCloud are priced in two categories:

**Base credits**

- Base credits are charged to ensure high availability for your application by maintaining at least one instance running continuously, ready to handle incoming requests. If you wish to stop the serving application, you can either remove the app completely or put it on pause, the latter allows you to resume the app serving based on persisted configurations (refer to [`fastapi-serve` CLI section](#-fastapi-serve-cli) for more information). Both options will halt the consumption of credits.
- Actual credits charged for base credits are calculated based on the [instance type as defined by Jina AI Cloud](https://docs.jina.ai/concepts/jcloud/configuration/#cpu-tiers).
- By default, instance type `C3` is used with a minimum of 1 instance and [Amazon EFS](https://aws.amazon.com/efs/) disk of size 1G, which means that if your application is served on JCloud, you will be charged ~10 credits per hour.
- You can change the instance type and the minimum number of instances by providing a YAML configuration file using the `--config` option. For example, if you want to use instance type `C4` with a minimum of 0 replicas, and 2G EFS disk, you can provide the following configuration file:
  ```yaml
  instance: C4
  autoscale_min: 0
  disk_size: 2G
  ```

**Serving credits**

- Serving credits are charged when your application is actively serving incoming requests.
- Actual credits charged for serving credits are calculated based on the credits for the instance type multiplied by the duration for which your application serves requests. 
- You are charged for each second your application is serving requests.


**Total credits charged = Base credits + Serving credits**. ([Jina AI Cloud](https://cloud.jina.ai/pricing) defines each credit as €0.005)

### Examples

<details>
<summary><b>Example 1</b></summary>

Consider an HTTP application that has served requests for `10` minutes in the last hour and uses a custom config:
```
instance: C4
autoscale_min: 0
disk_size: 2G
```

Total credits per hour charged would be `3.33`. The calculation is as follows:
```
C4 instance has an hourly credit rate of 20.
EFS has hourly credit rate of 0.104 per GB.
Base credits = 0 + 2 * 0.104 = 0.208 (since `autoscale_min` is 0)
Serving credits = 20 * 10/60 = 3.33
Total credits per hour = 0.208 + 3.33 = 3.538
```

</details>


<details>
<summary><b>Example 2</b></summary>

Consider a WebSocket application that had active connections for 20 minutes in the last hour and uses the default configuration.
```
instance: C3
autoscale_min: 1
disk_size: 1G
```

Total credits per hour charged would be `13.33`. The calculation is as follows:
```
C3 instance has an hourly credit rate of 10.
EFS has hourly credit rate of 0.104 per GB.
Base credits = 10 + 1 * 0.104 = 10.104 (since `autoscale_min` is 1)
Serving credits = 10 * 20/60 = 3.33
Total credits per hour = 10.104 + 3.33 = 13.434
```

</details>

CHANGELOG.md








<a name=release-note-0-0-1></a>
## Release Note (`0.0.1`)

> Release time: 2023-07-11 14:52:37



🙇 We'd like to thank all contributors for this new release! In particular,
 Deepankar Mahapatro,  🙇


### 📗 Documentation

 - [[```f8a0ef48```](https://github.com/jina-ai/langchain-serve/commit/f8a0ef483f25b399c9f947738fcae443ff703bfb)] __-__ update readme (*Deepankar Mahapatro*)

<a name=release-note-0-0-2></a>
## Release Note (`0.0.2`)

> Release time: 2023-07-14 12:08:49



🙇 We'd like to thank all contributors for this new release! In particular,
 Deepankar Mahapatro,  Jina Dev Bot,  🙇


### 🆕 New Features

 - [[```0f5b787a```](https://github.com/jina-ai/langchain-serve/commit/0f5b787a7018a6901eb89e68e3f99bc5ad21321e)] __-__ authorization with jina (*Deepankar Mahapatro*)
 - [[```cffe0d1d```](https://github.com/jina-ai/langchain-serve/commit/cffe0d1d39208e88d77c9cc18498a67f537f7db2)] __-__ cloud deployment (*Deepankar Mahapatro*)
 - [[```1d4ef64a```](https://github.com/jina-ai/langchain-serve/commit/1d4ef64a5b65629d5d1d47bffaa1b3c98b6dd78e)] __-__ fastapi gateway (*Deepankar Mahapatro*)
 - [[```127f46e8```](https://github.com/jina-ai/langchain-serve/commit/127f46e89953e08f56393c4f64c05025ecea1251)] __-__ repo, cicd setup (*Deepankar Mahapatro*)

### 📗 Documentation

 - [[```f3d560c1```](https://github.com/jina-ai/langchain-serve/commit/f3d560c108350b9677c8eae4f18cff6304fbbaff)] __-__ rate limit example (*Deepankar Mahapatro*)

### 🍹 Other Improvements

 - [[```e599280c```](https://github.com/jina-ai/langchain-serve/commit/e599280c299bc8eef770ee2ac727c62b94aae4dd)] __-__ __version__: the next version will be 0.0.2 (*Jina Dev Bot*)

<a name=release-note-0-0-3></a>
## Release Note (`0.0.3`)

> Release time: 2023-07-18 16:33:12



🙇 We'd like to thank all contributors for this new release! In particular,
 Deepankar Mahapatro,  Jina Dev Bot,  🙇


### 🆕 New Features

 - [[```5808f0b8```](https://github.com/jina-ai/langchain-serve/commit/5808f0b83b641df67ee5746643f39f0ef0fe0aae)] __-__ support jcloud yml for config (*Deepankar Mahapatro*)
 - [[```a954f998```](https://github.com/jina-ai/langchain-serve/commit/a954f9985a29fe51b8676de34b4bde1ffdbecec7)] __-__ blob storage for files (*Deepankar Mahapatro*)

### 🐞 Bug fixes

 - [[```57e64d55```](https://github.com/jina-ai/langchain-serve/commit/57e64d559a74e1478afa84b51268dc9766121386)] __-__ jcloud config reading (*Deepankar Mahapatro*)
 - [[```5d33505f```](https://github.com/jina-ai/langchain-serve/commit/5d33505f92766c54dc3333cf0831b72232406535)] __-__ move timeout from cli to jcloud yaml (*Deepankar Mahapatro*)

### 📗 Documentation

 - [[```033a36b4```](https://github.com/jina-ai/langchain-serve/commit/033a36b42414f7b3b3c7cdb7f54f6a46be11793c)] __-__ autoscaling with cpu usage (*Deepankar Mahapatro*)
 - [[```f461dae4```](https://github.com/jina-ai/langchain-serve/commit/f461dae4f98e7a3ec4e2352931a0f6e17503e33c)] __-__ update readme (*Deepankar Mahapatro*)

### 🍹 Other Improvements

 - [[```0c43c0e9```](https://github.com/jina-ai/langchain-serve/commit/0c43c0e9b3595b99580a3a851476fbd52e44b92f)] __-__ add readme (*Deepankar Mahapatro*)
 - [[```88e90e65```](https://github.com/jina-ai/langchain-serve/commit/88e90e65dbb7e2a0d2e413610d90738ce95763cd)] __-__ __version__: the next version will be 0.0.3 (*Jina Dev Bot*)

<a name=release-note-0-0-4></a>
## Release Note (`0.0.4`)

> Release time: 2023-07-19 16:36:47



🙇 We'd like to thank all contributors for this new release! In particular,
 Deepankar Mahapatro,  Jina Dev Bot,  🙇


### 🆕 New Features

 - [[```c3eb8d9b```](https://github.com/jina-ai/langchain-serve/commit/c3eb8d9b1212dae68eb77d39ab620610ccbb4da8)] __-__ export to docker-compose and k8s (*Deepankar Mahapatro*)

### 🐞 Bug fixes

 - [[```7b12f9e3```](https://github.com/jina-ai/langchain-serve/commit/7b12f9e35d1609804f55ff8c5bcedae4e563f775)] __-__ jcloud config reading (*Deepankar Mahapatro*)

### 📗 Documentation

 - [[```a203baf8```](https://github.com/jina-ai/langchain-serve/commit/a203baf8abeb836964d8edd7b59c9589635e27e5)] __-__ update readme (*Deepankar Mahapatro*)
 - [[```b17cf071```](https://github.com/jina-ai/langchain-serve/commit/b17cf071f181511d46ba2b57241b6c544223bc7f)] __-__ env var support (*Deepankar Mahapatro*)
 - [[```753b85fd```](https://github.com/jina-ai/langchain-serve/commit/753b85fde708339b891defc66ad461236781bd1c)] __-__ serverless autoscaling based on rps (*Deepankar Mahapatro*)

### 🍹 Other Improvements

 - [[```4eb2d8f7```](https://github.com/jina-ai/langchain-serve/commit/4eb2d8f7209bc2c20ca29b7cf6ed6cabee733dbb)] __-__ __version__: the next version will be 0.0.4 (*Jina Dev Bot*)

<a name=release-note-0-0-5></a>
## Release Note (`0.0.5`)

> Release time: 2023-07-20 08:44:32



🙇 We'd like to thank all contributors for this new release! In particular,
 Deepankar Mahapatro,  Jina Dev Bot,  🙇


### 🆕 New Features

 - [[```d86db472```](https://github.com/jina-ai/langchain-serve/commit/d86db472b2501ef6cbe2525bd7f20d387d237956)] __-__ local deployment (*Deepankar Mahapatro*)

### 📗 Documentation

 - [[```d1b28074```](https://github.com/jina-ai/langchain-serve/commit/d1b28074a5e34a54c17e0364aaa2ab60f61e5ca2)] __-__ update readme (*Deepankar Mahapatro*)

### 🍹 Other Improvements

 - [[```9d86af24```](https://github.com/jina-ai/langchain-serve/commit/9d86af24e25ae5e029bef0491c547a28c8fb5158)] __-__ __version__: the next version will be 0.0.5 (*Jina Dev Bot*)

<a name=release-note-0-0-6></a>
## Release Note (`0.0.6`)

> Release time: 2023-07-20 15:02:49



🙇 We'd like to thank all contributors for this new release! In particular,
 Deepankar Mahapatro,  Jina Dev Bot,  🙇


### 🆕 New Features

 - [[```3f73da3a```](https://github.com/jina-ai/langchain-serve/commit/3f73da3af0ccac03ec1c3347387b59e210968f18)] __-__ zero downtime upgrades (*Deepankar Mahapatro*)

### 🍹 Other Improvements


fastapi_serve/cloud/config.py
import os
from dataclasses import dataclass, field
from typing import Dict

import click
import yaml

from fastapi_serve.cloud.errors import (
    InvalidAutoscaleMaxError,
    InvalidAutoscaleMinError,
    InvalidDiskSizeError,
    InvalidInstanceError,
)

# jcloud args
VALID_AUTOSCALE_METRICS = ['cpu', 'memory', 'rps']

# default values
APP_NAME = 'fastapi'
JINA_VERSION = '3.18.0'
DOCARRAY_VERSION = '0.21.0'
DEFAULT_TIMEOUT = 120
DEFAULT_LABEL = 'fastapi-serve'

# jcloud urls
APP_LOGS_URL = "[https://cloud.jina.ai/](https://cloud.jina.ai/user/flows?action=detail&id={app_id}&tab=logs)"
PRICING_URL = "****{cph}**** ([Read about pricing here](https://github.com/jina-ai/fastapi-serve/blob/main/docs/CONFIG.MD#-pricing))"


@dataclass
class Defaults:
    instance: str = 'C3'  # instance type
    autoscale_min: int = 1  # min number of replicas
    autoscale_max: int = 10  # max number of replicas
    autoscale_metric: str = 'cpu'  # cpu, memory, rps
    autoscale_cpu_target: int = 70  # 70% cpu usage
    autoscale_rps_target: int = 10  # 10 requests per second
    autoscale_stable_window: int = DEFAULT_TIMEOUT
    autoscale_revision_timeout: int = DEFAULT_TIMEOUT
    disk_size: str = '1G'


@dataclass
class AutoscaleConfig:
    min: int = Defaults.autoscale_min
    max: int = Defaults.autoscale_max
    metric: str = Defaults.autoscale_metric
    target: int = Defaults.autoscale_cpu_target
    stable_window: int = Defaults.autoscale_stable_window
    revision_timeout: int = Defaults.autoscale_revision_timeout

    def __post_init__(self):
        try:
            self.min = int(self.min)
            if self.min < 0:
                raise InvalidAutoscaleMinError(self.min)
        except ValueError:
            raise InvalidAutoscaleMinError(self.min)

        try:
            self.max = int(self.max)
            if self.max < 1:
                raise InvalidAutoscaleMaxError(self.max)
        except ValueError:
            raise InvalidAutoscaleMaxError(self.max)

        if self.min > self.max:
            raise InvalidAutoscaleMaxError(self.max)

        if self.target < 1:
            raise ValueError(
                f'Invalid autoscale target {self.target}. Must be greater than 1'
            )

        if self.metric not in VALID_AUTOSCALE_METRICS:
            raise ValueError(
                f'Invalid autoscale metric {self.metric}. Valid options are {VALID_AUTOSCALE_METRICS}'
            )

        if self.metric in ['cpu', 'memory'] and self.min == 0:
            print(f'Cannot autoscale to 0 for {self.metric}. Resetting it to 1')
            self.min = 1

    def to_dict(self) -> Dict:
        return {
            'autoscale': {
                'min': self.min,
                'max': self.max,
                'metric': self.metric,
                'target': self.target,
                'stable_window': self.stable_window,
                'revision_timeout': self.revision_timeout,
            }
        }

    @classmethod
    def from_dict(cls, config: Dict):
        _autoscale = config.get('autoscale', {})
        return cls(
            min=_autoscale.get('min', cls.min),
            max=_autoscale.get('max', cls.max),
            metric=_autoscale.get('metric', cls.metric),
            target=_autoscale.get('target', cls.target),
        )


@dataclass
class JCloudConfig:
    timeout: int = DEFAULT_TIMEOUT
    instance: str = Defaults.instance
    disk_size: str = Defaults.disk_size
    autoscale: AutoscaleConfig = field(init=False)

    def __post_init__(self):
        if self.instance and not (
            self.instance.startswith(("C", "G")) and self.instance[1:].isdigit()
        ):
            raise InvalidInstanceError(self.instance)

        if self.disk_size is not None:
            if (
                isinstance(self.disk_size, str)
                and not self.disk_size.endswith(("M", "MB", "Mi", "G", "GB", "Gi"))
            ) or (isinstance(self.disk_size, int) and self.disk_size != 0):
                raise InvalidDiskSizeError(self.disk_size)

        if self.timeout < 1:
            raise ValueError(f'Invalid timeout {self.timeout}. Must be greater than 1')

        self.autoscale = AutoscaleConfig(
            stable_window=self.timeout, revision_timeout=self.timeout
        )

    def to_dict(self) -> Dict:
        jcloud_dict = {
            'jcloud': {
                'expose': True,
                'resources': {
                    'instance': self.instance,
                    'capacity': 'spot',
                },
                'network': {'healthcheck': False},
                'timeout': self.timeout,
                **self.autoscale.to_dict(),
            }
        }

        # Don't add the volume block if self.disk_size is 0
        if isinstance(self.disk_size, str):
            jcloud_dict['jcloud']['resources']['storage'] = {
                'kind': 'efs',
                'size': self.disk_size,
            }

        return jcloud_dict

    @classmethod
    def from_dict(cls, config: Dict):
        '''Sample config
        {
            "instance": "C3",
            "timeout": 120,
            "autoscale": {
                "min": 1,
                "max": 10,
                "metric": "cpu",
                "target": 70,
            },
            "disk_size": "1G"
        }
        '''
        jcloud_config = cls(
            instance=config.get('instance', cls.instance),
            disk_size=config.get('disk_size', cls.disk_size),
            timeout=config.get('timeout', cls.timeout),
        )
        jcloud_config.autoscale = AutoscaleConfig.from_dict(config)
        return jcloud_config

    @classmethod
    def from_file(cls, config_path: str):
        with open(config_path, "r") as f:
            config_data: Dict = yaml.safe_load(f)
            return cls.from_dict(config_data)


def validate_jcloud_config_callback(ctx, param, value):
    if not value:
        return None
    try:
        JCloudConfig.from_file(value)
    except InvalidInstanceError as e:
        raise click.BadParameter(
            f"Invalid instance '{e.instance}' found in config file', please refer to https://docs.jina.ai/concepts/jcloud/configuration/#cpu-tiers for instance definition."
        )
    except InvalidAutoscaleMinError as e:
        raise click.BadParameter(
            f"Invalid instance '{e.min}' found in config file', it should be a number >= 0."
        )
    except InvalidDiskSizeError as e:
        raise click.BadParameter(
            f"Invalid disk size '{e.disk_size}' found in config file."
        )

    return value


def resolve_jcloud_config(config, app_dir: str):
    # config given from CLI takes higher priority
    if config:
        return config

    # Check to see if jcloud YAML/YML file exists at app dir
    config_path_yml = os.path.join(app_dir, "jcloud.yml")
    config_path_yaml = os.path.join(app_dir, "jcloud.yaml")

    if os.path.exists(config_path_yml):
        config_path = config_path_yml
    elif os.path.exists(config_path_yaml):
        config_path = config_path_yaml
    else:
        return None

    try:
        JCloudConfig.from_file(config_path)
    except (InvalidAutoscaleMinError, InvalidInstanceError, InvalidDiskSizeError) as e:
        # If it's malformed, we treated as non-existed
        print(f'config file {config_path} is malformed: {e}')
        return None

    print(f'JCloud config file at app directory will be applied: {config_path}')
    return config_path


def get_jcloud_config(config_path: str = None) -> JCloudConfig:
    if not config_path:
        return JCloudConfig()

    if not os.path.exists(config_path):
        print(f'config file {config_path} not found')
        return JCloudConfig()

    with open(config_path, 'r') as f:
        config_data: Dict = yaml.safe_load(f)
        if not config_data:
            return JCloudConfig()

        return JCloudConfig.from_dict(config_data)
